vsphere/virtual_machine_config_structure.go

```diff
@@ -0,0 +1,664 @@
+		}
+	}
+
+	// Look for new values, in addition to changed values.
+	for k1, v1 := range new.(map[string]interface{}) {
+		var found bool
+		for k2, v2 := range old.(map[string]interface{}) {
+			if k1 == k2 {
+				found = true
+				if v1 != v2 {
+					// Value has changed, add it to the changeset
+					ov := &types.OptionValue{
+						Key:   k1,
+						Value: types.AnyType(v1),
+					}
+					opts = append(opts, ov)
+				}
+			}
+		}
+		if !found {
+			// Brand new value
+			ov := &types.OptionValue{
+				Key:   k1,
+				Value: types.AnyType(v1),
+			}
+			opts = append(opts, ov)
+		}
+	}
+
+	// Done!
+	return opts
+}
+
+// flattenExtraConfig reads in the extraConfig from a running virtual machine
+// and *only* sets the keys in extra_config that we know about. This is to
+// prevent Terraform from interfering with values that are maintained
+// out-of-band by vSphere which could lead to spurious diffs and unstable
+// operation.  Note the side-effect here is that Terraform cannot track manual
+// drift that is not a part of normal vSphere operation. Removing keys that
+// have been in configuration through at least one successful apply though are
+// safe, as removing them will add a nil value for that key in the next
+// chnageset, properly effecting its removal.
+func flattenExtraConfig(d *schema.ResourceData, opts []types.BaseOptionValue) error {
+	if len(opts) < 1 {
+		// No opts to read is a no-op
+		return nil
+	}
+	ec := make(map[string]interface{})
+	for _, v := range opts {
+		ov := v.GetOptionValue()
+		for k := range d.Get("extra_config").(map[string]interface{}) {
+			if ov.Key == k {
+				ec[ov.Key] = ov.Value
+			}
+		}
+	}
+	return d.Set("extra_config", ec)
+}
+
+// expandCPUCountConfig is a helper for expandVirtualMachineConfigSpec that
+// determines if we need to restart the VM due to a change in CPU count. This
+// is determined by the net change in CPU count and the pre-update values of
+// cpu_hot_add_enabled and cpu_hot_remove_enabled. The pre-update value is
+// important here as while CPU hot-add/remove is supported while the values are
+// enabled on the virtual machine, modification of hot-add/remove themselves is
+// an operation that requires a power down of the VM.
+func expandCPUCountConfig(d *schema.ResourceData) int32 {
+	occ, ncc := d.GetChange("num_cpus")
+	cha, _ := d.GetChange("cpu_hot_add_enabled")
+	currentHotAdd := cha.(bool)
+	chr, _ := d.GetChange("cpu_hot_remove_enabled")
+	currentHotRemove := chr.(bool)
+	oldCPUCount := int32(occ.(int))
+	newCPUCount := int32(ncc.(int))
+
+	switch {
+	case oldCPUCount < newCPUCount:
+		// Adding CPUs
+		if !currentHotAdd {
+			log.Printf("[DEBUG] %s: CPU operation requires a VM restart", resourceVSphereVirtualMachineIDString(d))
+			d.Set("reboot_required", true)
+		}
+	case oldCPUCount > newCPUCount:
+		// Removing CPUs
+		if !currentHotRemove {
+			log.Printf("[DEBUG] %s: CPU operation requires a VM restart", resourceVSphereVirtualMachineIDString(d))
```

Comment:
(as above) I think this log message wants `virtual machine %q` to be clearer?

Reply:
https://github.com/terraform-providers/terraform-provider-vsphere/pull/244#discussion_r153990392